File: utils/timeframes.py
"""
Shared Timeframe Constants for Backend

This module provides a single source of truth for all timeframe definitions.
Import from this module instead of defining timeframes locally.
"""

from typing import Dict, List, Tuple, Optional
from dataclasses import dataclass
from enum import Enum


class TimeframeUnit(Enum):
    """Timeframe unit types"""
    MINUTE = "M"
    HOUR = "H"
    DAY = "D"
    WEEK = "W"
    MONTH = "MN"


@dataclass
class TimeframeOption:
    """Timeframe option with metadata"""
    value: str
    label: str
    minutes: int
    unit: TimeframeUnit


# Standard timeframes with full metadata
TIMEFRAMES: List[TimeframeOption] = [
    TimeframeOption("M1", "1 Minute", 1, TimeframeUnit.MINUTE),
    TimeframeOption("M2", "2 Minutes", 2, TimeframeUnit.MINUTE),
    TimeframeOption("M3", "3 Minutes", 3, TimeframeUnit.MINUTE),
    TimeframeOption("M5", "5 Minutes", 5, TimeframeUnit.MINUTE),
    TimeframeOption("M10", "10 Minutes", 10, TimeframeUnit.MINUTE),
    TimeframeOption("M15", "15 Minutes", 15, TimeframeUnit.MINUTE),
    TimeframeOption("M30", "30 Minutes", 30, TimeframeUnit.MINUTE),
    TimeframeOption("H1", "1 Hour", 60, TimeframeUnit.HOUR),
    TimeframeOption("H2", "2 Hours", 120, TimeframeUnit.HOUR),
    TimeframeOption("H4", "4 Hours", 240, TimeframeUnit.HOUR),
    TimeframeOption("H6", "6 Hours", 360, TimeframeUnit.HOUR),
    TimeframeOption("H8", "8 Hours", 480, TimeframeUnit.HOUR),
    TimeframeOption("H12", "12 Hours", 720, TimeframeUnit.HOUR),
    TimeframeOption("D1", "1 Day", 1440, TimeframeUnit.DAY),
    TimeframeOption("D3", "3 Days", 4320, TimeframeUnit.DAY),
    TimeframeOption("W1", "1 Week", 10080, TimeframeUnit.WEEK),
    TimeframeOption("MN", "1 Month", 43200, TimeframeUnit.MONTH),
]

# Array of timeframe values only
TIMEFRAME_VALUES: List[str] = [tf.value for tf in TIMEFRAMES]

# Dictionary mapping value to label
TIMEFRAME_LABELS: Dict[str, str] = {tf.value: tf.label for tf in TIMEFRAMES}

# Dictionary mapping value to minutes
TIMEFRAME_MINUTES: Dict[str, int] = {tf.value: tf.minutes for tf in TIMEFRAMES}

# Common timeframes for quick access
class CommonTimeframes:
    ONE_MINUTE = "M1"
    FIVE_MINUTES = "M5"
    FIFTEEN_MINUTES = "M15"
    ONE_HOUR = "H1"
    FOUR_HOURS = "H4"
    ONE_DAY = "D1"
    ONE_WEEK = "W1"

# Short timeframe list for simple selectors
SHORT_TIMEFRAMES = ["M1", "M5", "M15", "H1", "H4", "D1", "W1"]

# Intraday timeframes only
INTRADAY_TIMEFRAMES = [tf for tf in TIMEFRAMES if tf.minutes < 1440]

# Swing trading timeframes
SWING_TIMEFRAMES = [tf for tf in TIMEFRAMES if 1440 <= tf.minutes < 10080]


def get_timeframe_by_value(value: str) -> Optional[TimeframeOption]:
    """Get timeframe option by value"""
    for tf in TIMEFRAMES:
        if tf.value == value:
            return tf
    return None


def get_timeframe_minutes(value: str) -> int:
    """Get minutes for a timeframe value, defaults to 60 (H1) if not found"""
    return TIMEFRAME_MINUTES.get(value, 60)


def validate_timeframe(value: str) -> bool:
    """Validate if a timeframe value is valid"""
    return value in TIMEFRAME_VALUES

File: frontend/src/lib/constants/timeframes.ts
/**
 * Shared Timeframe Constants
 * 
 * This file provides a single source of truth for all timeframe definitions.
 * Import from this file instead of defining timeframes locally.
 */

export interface TimeframeOption {
  value: string;
  label: string;
  minutes?: number;
}

/**
 * Standard timeframes with labels for UI display
 */
export const TIMEFRAMES: TimeframeOption[] = [
  { value: "M1", label: "1 Minute", minutes: 1 },
  { value: "M2", label: "2 Minutes", minutes: 2 },
  { value: "M3", label: "3 Minutes", minutes: 3 },
  { value: "M5", label: "5 Minutes", minutes: 5 },
  { value: "M10", label: "10 Minutes", minutes: 10 },
  { value: "M15", label: "15 Minutes", minutes: 15 },
  { value: "M30", label: "30 Minutes", minutes: 30 },
  { value: "H1", label: "1 Hour", minutes: 60 },
  { value: "H2", label: "2 Hours", minutes: 120 },
  { value: "H4", label: "4 Hours", minutes: 240 },
  { value: "H6", label: "6 Hours", minutes: 360 },
  { value: "H8", label: "8 Hours", minutes: 480 },
  { value: "H12", label: "12 Hours", minutes: 720 },
  { value: "D1", label: "1 Day", minutes: 1440 },
  { value: "D3", label: "3 Days", minutes: 4320 },
  { value: "W1", label: "1 Week", minutes: 10080 },
  { value: "MN", label: "1 Month", minutes: 43200 },
];

/**
 * Array of timeframe values only (for backward compatibility)
 */
export const TIMEFRAME_VALUES = TIMEFRAMES.map(tf => tf.value);

/**
 * Map of timeframe value to label
 */
export const TIMEFRAME_LABELS = Object.fromEntries(
  TIMEFRAMES.map(tf => [tf.value, tf.label])
) as Record<string, string>;

/**
 * Map of timeframe value to minutes
 */
export const TIMEFRAME_MINUTES = Object.fromEntries(
  TIMEFRAMES.filter(tf => tf.minutes !== undefined).map(tf => [tf.value, tf.minutes])
) as Record<string, number>;

/**
 * Common timeframes for quick access
 */
export const COMMON_TIMEFRAMES = {
  ONE_MINUTE: "M1",
  FIVE_MINUTES: "M5",
  FIFTEEN_MINUTES: "M15",
  ONE_HOUR: "H1",
  FOUR_HOURS: "H4",
  ONE_DAY: "D1",
  ONE_WEEK: "W1",
} as const;

/**
 * Short timeframe list for simple selectors
 */
export const SHORT_TIMEFRAMES = ["M1", "M5", "M15", "H1", "H4", "D1", "W1"];

/**
 * Intraday timeframes only
 */
export const INTRADAY_TIMEFRAMES = TIMEFRAMES.filter(tf => 
  tf.minutes !== undefined && tf.minutes < 1440
);

/**
 * Swing trading timeframes
 */
export const SWING_TIMEFRAMES = TIMEFRAMES.filter(tf => 
  tf.minutes !== undefined && tf.minutes >= 1440 && tf.minutes < 10080
);

export default TIMEFRAMES;
